### PR TITLE
table: check and truncate UTF8 string in CastValue.

### PR DESCRIPTION
**Pull request by coocood**
Fixes issue https://github.com/pingcap/tidb/issues/2817

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala PTAL

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -344,3 +344,18 @@ func (sc *StatementContext) AppendWarning(warn error) {
 	}
 	sc.mu.Unlock()
 }
+
+// HandleTruncate ignore or return the error based on the StatementContext state.
```
Comment (zimulala) [inline sessionctx/variable/session.go]: s/ignore/ignores
s/return/returns

File: table/table.go [hunk 2]
```diff
@@ -137,10 +139,11 @@ const (
...
 
-	codeColumnCantNull  = 1048
-	codeUnknownColumn   = 1054
-	codeDuplicateColumn = 1110
-	codeNoDefaultValue  = 1364
+	codeColumnCantNull     = 1048
+	codeUnknownColumn      = 1054
+	codeDuplicateColumn    = 1110
+	codeNoDefaultValue     = 1364
+	codeTruncateWrongValue = 1366
```
Comment (zimulala) [inline table/table.go]: Add it to tableMySQLErrCodes.

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -344,3 +344,18 @@ func (sc *StatementContext) AppendWarning(warn error) {
...
+		return nil
+	}
+	if sc.IgnoreTruncate {
+		return nil
+	}
+	if sc.TruncateAsWarning {
+		sc.AppendWarning(err)
+		return nil
+	}
+	return err
```
Comment (zimulala) [inline sessionctx/variable/session.go]: Add the error trace.

File: executor/statement_context_test.go [hunk 4]
```diff
@@ -65,4 +67,14 @@ func (s *testSuite) TestStatementContext(c *C) {
...
+
+	// Test invalid UTF8
+	tk.MustExec("create table sc2 (a varchar(255))")
+	// Insert an invalid UTF8
+	tk.MustExec("insert sc2 values (unhex('4040ffff'))")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.WarningCount(), Greater, uint16(0))
+	tk.MustQuery("select * from sc2").Check(testkit.Rows(fmt.Sprintf("%v", []byte("@@"))))
+	tk.MustExec(strictModeSQL)
+	_, err = tk.Exec("insert sc2 values (unhex('4040ffff'))")
+	c.Assert(err, NotNil)
```
Comment (zimulala) [inline executor/statement_context_test.go]: Check the error code.

File: sessionctx/variable/session.go (hunk 3 above)
Comment (coocood) [inline sessionctx/variable/session.go]: This trace is redundant, intentionally not trace it here.

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -344,3 +344,18 @@ func (sc *StatementContext) AppendWarning(warn error) {
 	}
 	sc.mu.Unlock()
 }
+
+// HandleTruncate ignores or returns the error based on the StatementContext state.
+func (sc *StatementContext) HandleTruncate(err error) error {
+	if err == nil {
```
Comment (shenli) [inline sessionctx/variable/session.go]: Should we check if the err is truncate error?

Comment (coocood) [inline sessionctx/variable/session.go]: There may be many kinds of truncate error, check it here need to add many dependencies creates dependency cycle, so 
we just let caller make sure it's a truncate error.

**Comment by coocood** [comment]
@shenli @zimulala PTAL

**Comment by shenli** [comment]
LGTM

File: table/column.go [hunk 6]
```diff
@@ -113,15 +114,26 @@ func CastValues(ctx context.Context, rec []types.Datum, cols []*Column, ignoreEr
...
+	sc := ctx.GetSessionVars().StmtCtx
+	casted, err = val.ConvertTo(sc, &col.FieldType)
+	// TODO: make sure all truncate errors are handled by ConvertTo.
+	err = sc.HandleTruncate(err)
 	if err != nil {
-		if ctx.GetSessionVars().StrictSQLMode {
-			return casted, errors.Trace(err)
+		return casted, errors.Trace(err)
+	}
+	if col.Charset != mysql.UTF8Charset && col.Charset != mysql.UTF8MB4Charset {
```
Comment (zimulala) [inline table/column.go]: You can use `mysql.IsUTF8Charset` directly.

**Pull request merged**